rclcpp/multithread_executor/multithread_executor.cpp

```diff
@@ -0,0 +1,194 @@
+
+    subscription2 = this->create_subscription<std_msgs::msg::String>("topic",
+        rclcpp::QoS(10),
+        std::bind(&DualThreadedNode::subscriber2_cb,
+        this,
+        std::placeholders::_1),
+        sub2_opt);
+  }
+
+private:
+  /**
+   * Simple function for generating a timestamp
+   * Used for somewhat ineffectually demonstrating that the multithreading doesn't cripple performace
+   */
+  std::string timing_string()
+  {
+    time_t timer;
+    struct tm * tm_info;
+    struct tm tm_result;
+    char buffer[12];
+
+    time(&timer);
+    tm_info = localtime_r(&timer, &tm_result);
+
+    strftime(buffer, 12, "[%T]", tm_info);
+
+    return buffer;
+  }
+
+  /**
+   * A small convenience function for converting a thread ID to a string
+   */
+  std::string string_thread_id(std::thread::id id)
+  {
+    auto hashed = std::hash<std::thread::id>()(id);
+    return std::to_string(hashed);
+  }
+
+  /**
+   * Every time the Publisher publishes something, all subscribers to the topic get poked
+   * This function gets called when Subscriber1 is poked (due to the std::bind we used when defining it)
+   */
+  void subscriber1_cb(const std_msgs::msg::String::SharedPtr msg)
+  {
+    auto message_recieved_at = timing_string();
+
+    // Extract current thread
+    auto thread_string = "THREAD " + string_thread_id(std::this_thread::get_id());
+    thread_string += " => Heard '%s' at " + message_recieved_at;  // Prep display message
+    RCLCPP_INFO(this->get_logger(), thread_string, msg->data.c_str());
+  }
+
+  /**
+   * This function gets called when Subscriber2 is poked
+   * Since it's running on a separate thread than Subscriber 1, it will run at (more-or-less) the same time!
+   */
+  void subscriber2_cb(const std_msgs::msg::String::SharedPtr msg)
+  {
+    auto message_recieved_at = timing_string();
+
+    // Extract current thread
+    auto thread_string = "THREAD " + string_thread_id(std::this_thread::get_id());
+
+    // Prep display message
+    thread_string += " => Heard '%s' at " + message_recieved_at;
+    RCLCPP_INFO(this->get_logger(), thread_string, msg->data.c_str());
+  }
+
+  rclcpp::callback_group::CallbackGroup::SharedPtr callback_group_subscriber1_;
+  rclcpp::callback_group::CallbackGroup::SharedPtr callback_group_subscriber2_;
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription1;
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription2;
+  rclcpp::SubscriptionOptions sub1_opt;
+  rclcpp::SubscriptionOptions sub2_opt;
+};
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  // You MUST use the MultiThreadedExecutor to use, well, multiple threads
+  rclcpp::executors::MultiThreadedExecutor executor;
+  auto pubnode = std::make_shared<PublisherNode>();
+  auto subnode = std::make_shared<DualThreadedNode>();  // This contains BOTH subscriber nodes. They will still run on different threads
```

Comment:
Maybe clarify which parts of the node are being run on multiple threads?

I also think that "subscriber nodes" should be rephrased to "subscription callbacks" (since it's still just one node).

Reply:
I have taken your suggestion and renamed "subscriber nodes" to "subscription callbacks". Definitely makes more sense.

I was hoping that the constructor would make the two threads clear enough. Should I add a few more comments around `subscription1_` and `subscription2_` to make it completely unambiguous?